stream/src/main/java/org/apache/apex/malhar/stream/window/impl/WindowedOperatorImpl.java

```diff
@@ -0,0 +1,436 @@
+             slideBeginTimestamp -= slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+        for (long slideBeginTimestamp = beginTimestamp + slideBy;
+             slideBeginTimestamp >= timestamp && timestamp > slideBeginTimestamp + durationMillis;
+             slideBeginTimestamp += slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+      }
+    } else {
+      throw new IllegalStateException("Unexpected WindowOption");
+    }
+    return windows;
+  }
+
+  @Override
+  public boolean isTooLate(long timestamp)
+  {
+    Duration allowedLateness = windowOption.getAllowedLateness();
+    return timestamp < currentWatermark - allowedLateness.getMillis();
+  }
+
+  @Override
+  public void dropTuple(InputT input)
+  {
+    // do nothing
+  }
+
+  @Override
+  public void accumulateTuple(WindowedValue<InputT> tuple)
+  {
+    KeyT key = keyExtractor.f(tuple.value);
+    for (Window window : tuple.windows) {
+      // process each window
+      AccumT accum = dataStorage.get(window, key);
+      dataStorage.put(window, key, accumulation.accumulate(accum, tuple.value));
+    }
+  }
+
+  @Override
+  public void processWatermark(Watermark watermark)
+  {
+    currentWatermark = watermark.getTimestamp();
+    long horizon = currentWatermark - windowOption.getAllowedLateness().getMillis();
+    // purge window that are too late to accept any more input
+    dataStorage.removeUpTo(horizon);
+
+    for (Iterator<Map.Entry<Window, WindowState>> it = windowStateMap.entrySet().iterator(); it.hasNext(); ) {
+      Map.Entry<Window, WindowState> entry = it.next();
+      Window window = entry.getKey();
+      WindowState windowState = entry.getValue();
+      if (window.getBeginTimestamp() + window.getDurationMillis() < horizon) {
+        // discard this window because it's too late now
+        it.remove();
+      } else if (window.getBeginTimestamp() + window.getDurationMillis() < currentWatermark) {
+        // watermark has not arrived for this window before, marking this window late
+        if (windowState.watermarkArrivalTime == -1) {
+          windowState.watermarkArrivalTime = currentDerivedTimestamp;
+          if (triggerAtWatermark) {
+            // fire trigger at watermark if applicable
+            fireTrigger(window, windowState);
+          }
+        }
+      }
+    }
+
+    // TODO: propagate the watermark downstream
+    //output.emit(watermark);
+
+    // TODO: if join operation (multiple input ports), we need to keep track of the watermark on all ports, and then propagate the watermark only when a new watermark arrives with the smallest timestamp among all the ports
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (this.firstWindowMillis <= 0) {
+      this.firstWindowMillis = System.currentTimeMillis();
```

Comment:
Should eventually be the DAG start time. We need to add that in core in the future. Till then we could also derive it from first windowId.